components/roster/heroes/Heroes.tsx

```diff
@@ -0,0 +1,283 @@
+  const [heroes, dispatch] = useReducer(reducer, heroList);
+
+  const CustomSelectSubmenu = ({
+    hero,
+    options,
+    selected,
+    heroProperty,
+    label,
+    disabled = false,
+  }: SubSelectProps) => (
+    <DropdownMenuSub>
+      <DropdownMenuSubTrigger
+        className={`w-full ${disabled ? "opacity-50 cursor-not-allowed" : ""}`}
+        disabled={disabled}
+      >
+        <span>{label}</span>
+      </DropdownMenuSubTrigger>
+      {!disabled && (
+        <DropdownMenuSubContent className="w-48">
+          <DropdownMenuLabel>{label}</DropdownMenuLabel>
+          <DropdownMenuSeparator />
+
+          <DropdownMenuSeparator />
+          {options.map((option) => (
+            <DropdownMenuItem
+              key={option}
+              onSelect={() =>
+                dispatch({ property: heroProperty, value: option, hero: hero })
+              }
+            >
+              <Check
+                className={`mr-2 h-4 w-4 ${
+                  selected === option ? "opacity-100" : "opacity-0"
+                }`}
+              />
+              {option}
+            </DropdownMenuItem>
+          ))}
+        </DropdownMenuSubContent>
+      )}
+    </DropdownMenuSub>
+  );
+
+  const handleSave = async (e: FormEvent) => {
+    e.preventDefault();
+
+    let items = heroes
+      .filter((hero) => hero.unlocked)
+      .map((hero) => ({
+        heroId: parseInt(hero.key),
+        ascension: hero.ascension,
+        equipment: hero.exEquipment,
+      }));
+
+    try {
+      const response = await (
+        await fetch("/api/roster/heroes", {
+          method: "POST",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify(items),
+        })
+      ).json();
+
+      toast.success("Heroes saved!");
+    } catch (error: any) {
+      toast.error("Failed to save Heroes!");
+    }
+  };
+
+  return (
+    <div className="flex flex-col items-center justify-center gap-2">
+      <div className="flex flex-row flex-wrap gap-x-2 items-center font-bold text-lg text-atekgold">
+        <span>Heroes</span>
+        <Button variant="analytica" onClick={handleSave}>
+          Save
+        </Button>
+      </div>
+      <div className="grid grid-cols-4 md:grid-cols-6 lg:grid-cols-10 gap-x-2 gap-y-6 pb-8">
+        {heroes.map((hero) => (
+          <div key={hero.key} className="cursor-pointer">
+            <DropdownMenu>
+              <DropdownMenuTrigger asChild>
+                <Card className="relative border-0 bg-slate-900">
+                  <CardContent className="flex flex-col justify-center gap-x-2 p-0">
+                    <div className="z-20">
+                      <HeroPortrait hero={hero} />
+                    </div>
+                    {isAscensionLevelValid(hero.ascension) && (
+                      <div className="z-10 absolute -bottom-4 w-[86%] h-full bg-slate-700 rounded-b-md mx-auto left-0 right-0 px-1">
```

Comment:
a `%` width here is going to be weird across devices. why is it needed? if you do `w-full` with x padding wouldn't that give the same result?

Reply:
No it doesn't. Also this 86% is consistent because it's always 76% of it's current container because it's absolute and it's parent is relative. I see what you're saying though, but nah that doesn't work if I do `w-full` and then put addition x padding on it. This is where I'm not CSS expert ha.